mava/systems/jax/config.py

```diff
@@ -0,0 +1,153 @@
+                    if name in list(self._config.keys()):
+                        raise Exception(
+                            "The given component config is already part of the current \
+                            system. Perhaps try updating the component instead using \
+                            .update() in the system builder."
+                        )
+                    else:
+                        new_param_names = list(dataclass.__dict__.keys())
+                        if set(self._current_params) & set(new_param_names):
+                            raise Exception(
+                                "Component configs share a common parameter name. \
+                                This is not allowed, please ensure config \
+                                parameter names are unique."
+                            )
+                        else:
+                            self._current_params.extend(new_param_names)
+                        self._config[name] = dataclass
+                else:
+                    raise Exception("Component configs must be a dataclass.")
+        else:
+            raise Exception(
+                "Component configs cannot be added to an already built config."
+            )
+
+    def update(self, **kwargs: Any) -> None:
+        """Update a component config dataclass.
+
+        Raises:
+            Exception: if a config shares a parameter name with another config
+            Exception: if a config is not already part of the system
+            Exception: if a config is not a dataclass object
+        """
+        if not self._built:
+            for name, dataclass in kwargs.items():
+                if is_dataclass(dataclass):
+                    if name in list(self._config.keys()):
+                        new_param_names = list(dataclass.__dict__.keys())
+                        if set(self._current_params) & set(new_param_names):
+                            raise Exception(
+                                "Component configs share a common parameter name. \
+                                This is not allowed, please ensure config \
+                                parameter names are unique."
+                            )
+                        else:
+                            self._current_params.extend(new_param_names)
+                        self._config[name] = dataclass
+                    else:
+                        raise Exception(
+                            "The given component config is not part of the current \
+                            system. Perhaps try adding the component using .add() \
+                            in the system builder."
+                        )
+                else:
+                    raise Exception("Component configs must be a dataclass.")
+        else:
+            raise Exception(
+                "Component configs cannot be updated if config has already been built."
+            )
+
+    def build(self) -> None:
```

Comment:
Why do we have a build and a launch? Can't we just build it when we launch it? Then we don't have to check if it has already been built?

Reply:
There is no launch in the config? The building here separates the process of configs being added and updated when the system components are being added and updated and the process of setting the system hyperparameters. Once all the components have been added updated, the system can be configured, this then has the config built internally by the system which allows the user to set the parameters of the system before launching. Perhaps it might help to look at the acceptance test (#418) and general system class (#425).

Reply:
There is no launch in the config? The building here separates the process of configs being added and updated when the system components are being added and updated and the process of setting the system hyperparameters. Once all the components have been added updated, the system can be configured, this then has the config built internally by the system which allows the user to set the parameters of the system before launching. Perhaps it might help to look at the acceptance test (#418) and general system class (#425).